File: src/neetcode/Java/_22GenerateParentheses.java
package neetcode.Java;

import java.util.ArrayList;
import java.util.List;

public class _22GenerateParentheses {
    List<String> list = new ArrayList<>();

    public List<String> generateParenthesis(int n) {
        Helper("", n, n);
        return list;
    }

    public void Helper(String s, int open, int close) {
        if (open == 0 && close == 0) {
            list.add(s);
        }
        if (open > 0)
            Helper(s + "(", open - 1, close);
        if (close > open)
            Helper(s + ")", open, close - 1);
    }

    public List<String> generateParenthesis1(int n) {
        List<String> list = new ArrayList<>();
        Helper("", n, list, 0, 0);
        return list;
    }

    public void Helper(String s, int n, List<String> list, int np, int nc) {
        if (np < n) {
            Helper(s + "(", n, list, np + 1, nc);
        }
        if (nc < np) {
            Helper(s + ")", n, list, np, nc + 1);
        }
        if (np == nc && np == n) {
            list.add(s);
        }
    }


    public static void main(String[] args) {
        _22GenerateParentheses obj = new _22GenerateParentheses();
        System.out.println(obj.generateParenthesis(3));
        System.out.println(obj.generateParenthesis1(3));
    }

}

File: src/neetcode/kotlin/_22GenerateParentheses.kt
package neetcode.kotlin

class _22GenerateParentheses {
    var list: MutableList<String> = ArrayList()

    fun generateParenthesis(n: Int): List<String> {
        Helper("", n, n)
        return list
    }

    fun Helper(s: String, open: Int, close: Int) {
        if (open == 0 && close == 0) {
            list.add(s)
        }
        if (open > 0) Helper("$s(", open - 1, close)
        if (close > open) Helper("$s)", open, close - 1)
    }

    fun generateParenthesis1(n: Int): List<String> {
        val list: MutableList<String> = ArrayList()
        Helper("", n, list, 0, 0)
        return list
    }

    fun Helper(s: String, n: Int, list: MutableList<String>, np: Int, nc: Int) {
        if (np < n) {
            Helper("$s(", n, list, np + 1, nc)
        }
        if (nc < np) {
            Helper("$s)", n, list, np, nc + 1)
        }
        if (np == nc && np == n) {
            list.add(s)
        }
    }


    companion object {
        @JvmStatic
        fun main(args: Array<String>) {
            val obj = _22GenerateParentheses()
            println(obj.generateParenthesis(3))
            println(obj.generateParenthesis1(3))
        }
    }
}
